### PR TITLE
Implement `foreach` with `..< `

### LINKED ISSUE #141: Implement `foreach` with `..<`
Implement a subset of `foreach` that uses `..<` e.g.

```
foreach int i in 0 ..< v.length() {
   //...
}
```
This can turn into the same as a `while`, but for incrementing the loop variable, we can use an addition without the need to check for overflow.

This is useful with #130 or #135

### PR DESCRIPTION
**Pull request by manuranga**
Fixes  #141

### CODE REVIEW

**Comment by manuranga** [comment]
Integer addition optimization is not handled here and moved to #180.

File: compiler/modules/front/ast.bal [hunk 1]
```diff
@@ -28,9 +28,9 @@ type FunctionDef record {|
     bir:FunctionSignature? signature = ();
 |};
 
-type Stmt VarDeclStmt|AssignStmt|CallStmt|ReturnStmt|IfElseStmt|WhileStmt|BreakStmt|ContinueStmt;
+type Stmt VarDeclStmt|AssignStmt|CallStmt|ReturnStmt|IfElseStmt|WhileStmt|ForeachStmt|BreakStmt|ContinueStmt;
 type CallStmt FunctionCallExpr|MethodCallExpr;
-type Expr SimpleConstExpr|BinaryExpr|UnaryExpr|FunctionCallExpr|MethodCallExpr|VarRefExpr|TypeCastExpr|ConstructorExpr|MemberAccessExpr;
+type Expr SimpleConstExpr|BinaryExpr|UnaryExpr|FunctionCallExpr|MethodCallExpr|VarRefExpr|TypeCastExpr|ConstructorExpr|MemberAccessExpr|RangeExpr;
```
Comment (jclark) [inline compiler/modules/front/ast.bal:33]: In this subset , we don't allow an Expr to be a `RangeExpr`, so this should not be here.

File: compiler/modules/front/ast.bal [hunk 2]
```diff
@@ -57,6 +57,15 @@ type WhileStmt record {|
     Stmt[] body;
 |};
 
+type ForeachStmt record {|
+    InlineTypeDesc td;
```
Comment (jclark) [inline compiler/modules/front/ast.bal:61]: Don't need this: we only accept int.

File: compiler/modules/front/ast.bal [hunk 3]
```diff
@@ -148,6 +158,12 @@ type MemberAccessLExpr record {|
     err:Position pos;
 |};
 
+type RangeExpr record {|
+    Expr lower;
+    Expr upper;
+    RangeOp op;
```
Comment (jclark) [inline compiler/modules/front/ast.bal:164]: Don't need this: we only have "..<".

File: compiler/modules/front/astString.bal [hunk 4]
```diff
@@ -89,6 +89,14 @@ function stmtToWords(Word[] w, Stmt stmt) {
         exprToWords(w, stmt.condition);
         blockToWords(w, stmt.body);
     }
+    else if stmt is ForeachStmt{
```
Comment (jclark) [inline compiler/modules/front/astString.bal:92]: Missing space

File: compiler/modules/front/codeGen.bal [hunk 5]
```diff
@@ -159,6 +162,59 @@ function codeGenStmts(CodeGenContext cx, bir:BasicBlock bb, Scope? initialScope,
     return curBlock;
 }
 
+final readonly & map<bir:OrderOp> RANGE_TO_ORDER = {
```
Comment (jclark) [inline compiler/modules/front/codeGen.bal:165]: Don't need this: `...` is not in the subset.

File: compiler/modules/front/tests/testParser.bal [hunk 6]
```diff
@@ -250,6 +250,8 @@ function validTokenSourceFragments() returns string[][]|error {
          ["V", "expr", "a | b", "a | b"],
          ["V", "expr", "a & b", "a & b"],
          ["V", "expr", "a ^ b", "a ^ b"],
+         ["U", "expr", "a ..< b", "a ..< b"],
+         ["U", "expr", "a ... b", "a ... b"],
```
Comment (jclark) [inline compiler/modules/front/tests/testParser.bal:254]: Not in the subset.

File: compiler/modules/front/parseStmt.bal [hunk 7]
```diff
@@ -197,3 +201,17 @@ function parseWhileStmt(Tokenizer tok) returns WhileStmt|err:Syntax {
     Stmt[] body = check parseStmtBlock(tok);
     return { condition, body };
 }
+
+function parseForeachStmt(Tokenizer tok) returns ForeachStmt|err:Syntax {
+    InlineTypeDesc td = check parseInlineTypeDesc(tok);
```
Comment (jclark) [inline compiler/modules/front/parseStmt.bal:206]: The subset just accepts `int` here.

File: compiler/modules/front/codeGen.bal [hunk 8]
```diff
@@ -159,6 +162,59 @@ function codeGenStmts(CodeGenContext cx, bir:BasicBlock bb, Scope? initialScope,
...
+    "..<": "<"
+};
+
+function codeGenForeachStmt(CodeGenContext cx, bir:BasicBlock startBlock, Scope? scope, ForeachStmt stmt) returns CodeGenError|bir:BasicBlock? {
+    Expr range = stmt.iterable;
+    var { varName, semType} = stmt;
+    if lookup(varName, scope) !== () {
+        return cx.semanticErr(`duplicate declaration of ${varName}`);
+    }
+    if !(range is RangeExpr) {
```
Comment (jclark) [inline compiler/modules/front/codeGen.bal:176]: The ForeachStmt should have a range member: then you don't need this extra level of indentation.

File: compiler/modules/front/codeGen.bal [hunk 9]
```diff
@@ -474,6 +530,10 @@ function codeGenExpr(CodeGenContext cx, bir:BasicBlock bb, Scope? scope, Expr ex
                 return cx.semanticErr("cannot apply member access to constant of simple type");
             }
         }
+        // Range
+        var {upper, lower, op} => {
```
Comment (jclark) [inline compiler/modules/front/codeGen.bal:534]: We don't need this if we don't include ranges in expressions.

File: compiler/modules/front/codeGen.bal [hunk 10]
```diff
@@ -607,7 +667,7 @@ function mustLookup(CodeGenContext cx, string name, Scope? scope, boolean forAss
         return cx.semanticErr(`variable ${name} not found`);
     }
     else if forAssign && binding.isFinal {
-        return cx.semanticErr(`cannot assign to parameter ${name}`);
+        return cx.semanticErr(`cannot assign to final variable ${name}`);
```
Comment (jclark) [inline compiler/modules/front/codeGen.bal:670]: How does this deal with the parameter case?

**Review by jclark — COMMENTED** [review]
You have done a bunch of things that make this more complicated than is needed for this subset.

I will fix them.

File: compiler/modules/front/codeGen.bal (hunk 10 above)
Comment (manuranga) [reply compiler/modules/front/codeGen.bal:670]: Both assignment to loop var and assignment to parameter come here. Currently way to distinguish.

**Pull request merged**